pkg/kubelet/dockershim/docker_container_windows_test.go

```diff
@@ -0,0 +1,308 @@
+		defer setRegistryCreateKeyFunc(t, &dummyRegistryKey{})()
+
+		createConfig := &dockertypes.ContainerCreateConfig{}
+		cleanupInfo := &containerCreationCleanupInfo{}
+		err := applyGMSAConfig(containerConfigWithGMSAAnnotation, createConfig, cleanupInfo)
+
+		assert.Nil(t, err)
+
+		if assert.NotNil(t, createConfig.HostConfig) && assert.Equal(t, 1, len(createConfig.HostConfig.SecurityOpt)) {
+			secOpt := createConfig.HostConfig.SecurityOpt[0]
+
+			expectedPrefix := "credentialspec=registry://k8s-cred-spec-"
+			assert.Equal(t, expectedPrefix, secOpt[:len(expectedPrefix)])
+
+			hex := secOpt[len(expectedPrefix):]
+			hexRegex := regexp.MustCompile("^[0-9a-f]{80}$")
+			assert.True(t, hexRegex.MatchString(hex))
+			assert.NotEqual(t, expectedHex, hex)
+
+			assert.Equal(t, "k8s-cred-spec-"+hex, cleanupInfo.gMSARegistryValueName)
+		}
+	})
+	t.Run("when there's an error generating the random value name", func(t *testing.T) {
+		defer setRandomReader([]byte{})()
+
+		err := applyGMSAConfig(containerConfigWithGMSAAnnotation, &dockertypes.ContainerCreateConfig{}, &containerCreationCleanupInfo{})
+
+		if assert.NotNil(t, err) {
+			assert.True(t, strings.Contains(err.Error(), "unable to generate random registry value name"))
+		}
+	})
+	t.Run("if there's an error opening the registry key", func(t *testing.T) {
+		defer setRegistryCreateKeyFunc(t, &dummyRegistryKey{}, fmt.Errorf("dummy error"))()
+
+		err := applyGMSAConfig(containerConfigWithGMSAAnnotation, &dockertypes.ContainerCreateConfig{}, &containerCreationCleanupInfo{})
+
+		if assert.NotNil(t, err) {
+			assert.True(t, strings.Contains(err.Error(), "unable to open registry key"))
+		}
+	})
+	t.Run("if there's an error writing to the registry key", func(t *testing.T) {
+		key := &dummyRegistryKey{}
+		key.setStringValueError = fmt.Errorf("dummy error")
+		defer setRegistryCreateKeyFunc(t, key)()
+
+		err := applyGMSAConfig(containerConfigWithGMSAAnnotation, &dockertypes.ContainerCreateConfig{}, &containerCreationCleanupInfo{})
+
+		if assert.NotNil(t, err) {
+			assert.True(t, strings.Contains(err.Error(), "unable to write into registry value"))
+		}
+		assert.True(t, key.closed)
+	})
+	t.Run("if there is no GMSA annotation", func(t *testing.T) {
+		createConfig := &dockertypes.ContainerCreateConfig{}
+
+		err := applyGMSAConfig(&runtimeapi.ContainerConfig{}, createConfig, &containerCreationCleanupInfo{})
+
+		assert.Nil(t, err)
+		assert.Nil(t, createConfig.HostConfig)
+	})
+}
+
+func TestRemoveGMSARegistryValue(t *testing.T) {
+	valueName := "k8s-cred-spec-1900254518529e2a3dedb85cdec03ce270559647459ab531f07af5eb1c5495fda709435ce82ab89c"
+	cleanupInfoWithValue := &containerCreationCleanupInfo{gMSARegistryValueName: valueName}
+
+	t.Run("it does remove the registry value", func(t *testing.T) {
+		key := &dummyRegistryKey{}
+		defer setRegistryCreateKeyFunc(t, key)()
+
+		err := removeGMSARegistryValue(cleanupInfoWithValue)
+
+		assert.Nil(t, err)
```

Comment:
use `require.Nil` if you want the test fail and not continue past this line

Reply:
I think the same applies to a lot of other unit tests.

Reply:
In this case though, it might still be worth continuing after failing the test? At least there doesn't seem to be any harm in it. I've taken care to carefully protect against panics in all of these tests (using `if assert...` to protect lines that could panic otherwise).

I'll review the rest of the tests with an eye for that though.

Reply:
(did change it in a number of other tests where the `if assert....` was only protecting the last test)